### ISSUE TITLE
DataBus changes in NServiceBus 8 do not work with NewtonsoftJsonSerializer

### ISSUE THREAD

**Issue by bording** [Particular/NServiceBus#6675]
The changes made to the DataBus in #6392 do not seem to have taken into account how the `NewtonsoftJsonSerializer` deserializes messages, resulting in the `IDataBusSerializer` instance getting a `null` value passed into `propertyType` when called from [`DataBusReceiveBehavior`](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/DataBus/DataBusReceiveBehavior.cs#L54).

When the default XML serializer is used, `dataBusProperty.Type` has the value of `byte[]`, but when using `NewtonsoftJsonSerializer` the type information is not set, so the `Type` property is `null`.

**Comment by ankitvijay** [comment]
HI @bording, I ran into the issue today. Is there an ETA for when the fix will be available or a workaround I can use in the meantime?

```
info: NServiceBus.ImmediateRetry[0]
      Immediate Retry is going to retry message '1ab2fb66-07b7-4c26-90e5-af9d009a3c44' because of an exception:
      System.ArgumentNullException: Value cannot be null. (Parameter 'returnType')
         at System.Text.Json.ThrowHelper.ThrowArgumentNullException(String parameterName)
         at System.Text.Json.JsonSerializer.Deserialize(Stream utf8Json, Type returnType, JsonSerializerOptions options)
         at NServiceBus.SystemJsonDataBusSerializer.Deserialize(Type propertyType, Stream stream) in /_/src/NServiceBus.Core/DataBus/SystemJsonDataBusSerializer.cs:line
 26
         at NServiceBus.DataBusDeserializer.Deserialize(String serializerUsed, Type propertyType, Stream stream) in /_/src/NServiceBus.Core/DataBus/DataBusDeserializer.
cs:line 53
         at NServiceBus.DataBusReceiveBehavior.Invoke(IIncomingLogicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/DataBus/DataBusReceiveBehavior.cs:
line 54
         at NServiceBus.MutateIncomingMessageBehavior.InvokeIncomingMessageMutators(IIncomingLogicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Mess
ageMutators/MutateInstanceMessage/MutateIncomingMessageBehavior.cs:line 61
         at NServiceBus.DeserializeMessageConnector.Invoke(IIncomingPhysicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/Deseriali
zeMessageConnector.cs:line 32
         at NServiceBus.UnitOfWorkBehavior.InvokeUnitsOfWork(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/UnitOfWork/UnitOfWorkBehav
ior.cs:line 42
         at NServiceBus.UnitOfWorkBehavior.InvokeUnitsOfWork(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/UnitOfWork/UnitOfWorkBehav
ior.cs:line 64
         at NServiceBus.MutateIncomingTransportMessageBehavior.InvokeIncomingTransportMessagesMutators(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/N
ServiceBus.Core/MessageMutators/MutateTransportMessage/MutateIncomingTransportMessageBehavior.cs:line 60
         at NServiceBus.ProcessingStatisticsBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Performance/Statistics/Proc
essingStatisticsBehavior.cs:line 25
         at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/Pipeline/Incoming/Tr
ansportReceiveToPhysicalMessageConnector.cs:line 35
         at NServiceBus.RetryAcknowledgementBehavior.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/ServicePlatform/Retries/RetryAckno
wledgementBehavior.cs:line 25
         at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipeline
Executor.cs:line 45
         at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipeline
Executor.cs:line 64
         at NServiceBus.Transport.AzureServiceBus.MessagePump.ProcessMessage(ServiceBusReceivedMessage message, ProcessMessageEventArgs processMessageEventArgs, String
messageId, Dictionary`2 headers, BinaryData body, CancellationToken messageProcessingCancellationToken) in /_/src/Transport/Receiving/MessagePump.cs:line 295

```

**Comment by bording** [comment]
> HI @bording, I ran into this issue today. Is there an ETA for when the fix will be available or a workaround I can use in the meantime?

I don't have any ETA to share at this time. This public issue will be the place to watch to see when we've fixed the problem.

We have triaged the issue internally, so it will be fixed according to that prioritization.

As far as workarounds go, unfortunately there is no workaround other than to go back to NServiceBus 7.

**Comment by ankitvijay** [comment]
Hi @bording,
Thanks for the update. I understand. Unfortunately,  it would mean that we would not able to use the `DataBus` feature since our entire downstream is `NewtonsoftJsonSerializer` and I would try to avoid downgrading the NServiceBus nuget package.

- Can we create the Newtonsoft equivalent of `SystemJsonDataBusSerializer`? Will it work?
- Or, I'm happy to contribute to this issue if that is possible and you could give me some pointers.

**Comment by bording** [comment]
> Thanks for the update. I understand. Unfortunately, it would mean that we would not able to use the DataBus feature since our entire downstream is NewtonsoftJsonSerializer and I would try to avoid downgrading the NServiceBus nuget package.

For your endpoints that need to use the DataBus, I would recommend holding off on moving them to NServiceBus 8 until we can fix this bug. It is fine to have some endpoints on 7 and some on 8.

> Can we create the Newtonsoft equivalent of SystemJsonDataBusSerializer? Will it work?

The DataBus serializer is not the source of the problem, so having a Newtonsoft option for the DataBus would not prevent the `null` that causing the issue.

> Or, I'm happy to contribute to this issue if that is possible and you could give me some pointers.

Unfortunately this bug is in the core of NServiceBus, so it's a rather complex issue and not something I could easily provide enough context for.

**Comment by jpalac** [comment]
Will be shipped with Release 8.0.6 and 8.1.4

**Comment by jpalac** [comment]
[NServiceBus 8.0.6](https://github.com/Particular/NServiceBus/releases/tag/8.0.6) and [NServiceBus 8.1.4](https://github.com/Particular/NServiceBus/releases/tag/8.1.4) have been released
